gpbft/participant.go

```diff
@@ -157,57 +159,128 @@ func (p *Participant) beginInstance() error {
+	if p.gpbft, err = newInstance(p, p.currentInstance, chain, *comm.power, comm.beacon); err != nil {
+		return fmt.Errorf("failed creating new gpbft instance: %w", err)
 	}
-	p.nextInstance += 1
-	if err := p.granite.Start(); err != nil {
-		return fmt.Errorf("failed starting granite instance: %w", err)
+	if err := p.gpbft.Start(); err != nil {
+		return fmt.Errorf("failed starting gpbft instance: %w", err)
 	}
 	// Deliver any queued messages for the new instance.
-	for _, msg := range p.mqueue[p.granite.instanceID] {
+	for _, msg := range p.mqueue.Drain(p.gpbft.instanceID) {
 		if p.terminated() {
 			break
 		}
 		if p.tracer != nil {
-			p.tracer.Log("Delivering queued P%d{%d} ← P%d: %v", p.id, p.granite.instanceID, msg.Sender, msg)
+			p.tracer.Log("Delivering queued P%d{%d} ← P%d: %v", p.id, p.gpbft.instanceID, msg.Sender, msg)
 		}
-		if err := p.granite.Receive(msg); err != nil {
+		if valid, err := p.gpbft.Receive(msg); !valid {
+			// Silently drop invalid messages.
+		} else if err != nil {
 			return fmt.Errorf("delivering queued message: %w", err)
 		}
 	}
-	delete(p.mqueue, p.granite.instanceID)
-	return p.handleDecision()
+	p.handleDecision()
+	return nil
+}
+
+func (p *Participant) getCommittee(instance uint64) (*committee, error) {
+	c, ok := p.committees[instance]
+	if !ok {
+		power, beacon, err := p.host.GetCommitteeForInstance(instance)
+		if err != nil {
+			return nil, fmt.Errorf("failed fetching committee for instance %d: %w", instance, err)
+		}
+		if err := power.Validate(); err != nil {
+			return nil, fmt.Errorf("invalid power table: %w", err)
+		}
+		c = &committee{power: power, beacon: beacon}
+		p.committees[instance] = c
+	}
+	return c, nil
 }
 
-func (p *Participant) handleDecision() error {
+func (p *Participant) handleDecision() {
 	if p.terminated() {
-		p.finalised = p.granite.terminationValue
-		p.terminatedDuringRound = p.granite.round
-		p.granite = nil
+		p.finalised = p.gpbft.terminationValue
+		p.terminatedDuringRound = p.gpbft.round
+		delete(p.committees, p.gpbft.instanceID)
+		p.gpbft = nil
+		p.currentInstance++
 		nextStart := p.host.ReceiveDecision(p.finalised)
 
 		// Set an alarm at which to fetch the next chain and begin a new instance.
 		p.host.SetAlarm(nextStart)
-		return nil
 	}
-	return nil
 }
 
 func (p *Participant) terminated() bool {
-	return p.granite != nil && p.granite.phase == TERMINATED_PHASE
+	return p.gpbft != nil && p.gpbft.phase == TERMINATED_PHASE
 }
 
 func (p *Participant) Describe() string {
-	if p.granite == nil {
+	if p.gpbft == nil {
 		return "nil"
 	}
-	return p.granite.Describe()
+	return p.gpbft.Describe()
+}
+
+// A power table and beacon value used as the committee inputs to an instance.
+type committee struct {
+	power  *PowerTable
+	beacon []byte
+}
+
+// A collection of messages queued for delivery for a future instance.
+// The queue drops equivocations and unjustified messages beyond some round number.
+type messageQueue struct {
```

Comment:
Might be worth keeping it sorted with a priority queue, similar to what's done for message queue in `sim` package. 

We can revisit this with benchmarks at a later date.

Reply:
The only inspection of this queue is when it's completely drained, so I expect a single sort at the end will be more efficient than doing so incrementally.